Examples/Io/Root/include/ActsExamples/Io/Root/RootSimHitReader.hpp

```diff
@@ -0,0 +1,111 @@
+
+#include <cstddef>
+#include <cstdint>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <utility>
+#include <vector>
+
+class TChain;
+
+namespace ActsExamples {
+struct AlgorithmContext;
+
+/// @class RootParticleReader
+///
+/// @brief Reads in Particles information from a root file
+class RootSimHitReader : public IReader {
+ public:
+  /// @brief The nested configuration struct
+  struct Config {
+    /// name of the whiteboard entry
+    std::string simHitCollection = "simhits";
+    /// name of the output tree
+    std::string treeName = "hits";
+    ///< The name of the input file
+    std::string filePath;
+    /// Whether the events are ordered or not
+    bool orderedEvents = true;
+  };
+
+  RootSimHitReader(const RootSimHitReader &) = delete;
+  RootSimHitReader(const RootSimHitReader &&) = delete;
+
+  /// Constructor
+  /// @param config The Configuration struct
+  RootSimHitReader(const Config &config, Acts::Logging::Level level);
+
+  /// Framework name() method
+  std::string name() const override { return "RootSimHitReader"; }
+
+  /// Return the available events range.
+  std::pair<size_t, size_t> availableEvents() const override;
+
+  /// Read out data from the input stream
+  ///
+  /// @param context The algorithm context
+  ProcessCode read(const ActsExamples::AlgorithmContext &context) override;
+
+  /// Readonly access to the config
+  const Config &config() const { return m_cfg; }
+
+ private:
+  /// Private access to the logging instance
+  const Acts::Logger &logger() const { return *m_logger; }
+
+  /// The config class
+  Config m_cfg;
+
+  WriteDataHandle<SimHitContainer> m_outputSimHits{this, "OutputSimHits"};
+  std::unique_ptr<const Acts::Logger> m_logger;
+
+  /// mutex used to protect multi-threaded reads
+  std::mutex m_read_mutex;
+
+  /// Vector of {eventNr, entryMin, entryMax}
+  std::vector<std::tuple<uint32_t, std::size_t, std::size_t>> m_eventMap;
+
+  /// The input tree name
+  TChain *m_inputChain = nullptr;
+
+  /// The entry numbers for accessing events in increased order (there could be
+  /// multiple entries corresponding to one event number)
+  std::vector<long long> m_entryNumbers = {};
+
+  /// The keys we have in the ROOT file
+  constexpr static std::array<const char *, 12> m_floatKeys = {
+      "tx",  "ty", "tz",      "tt",      "tpx",     "tpy",
+      "tpz", "te", "deltapx", "deltapy", "deltapz", "deltae"};
+  constexpr static std::array<const char *, 2> m_uint64Keys = {"geometry_id",
+                                                               "particle_id"};
+  constexpr static std::array<const char *, 6> m_uint32Keys = {
+      "event_id", "volume_id",   "boundary_id",
+      "layer_id", "approach_id", "sensitive_id"};
+  constexpr static std::array<const char *, 1> m_int32Keys = {"index"};
+
+  std::unordered_map<std::string_view, float> m_floatColumns;
+  std::unordered_map<std::string_view, unsigned long long> m_uint64Columns;
```

Comment:
I guess this type comes from the writer?

Reply:
yeah this is really weird. The writer uses `uint64_t`. But if I use `uint64_t` in the reader, this gives a type mismatch... because I think this  is internally somehow converted to `ULong_t` which is a `typedef` to `unsigend long long`. Weird ROOT stuff...

Reply:
but deserves a comment actually